adapters/adgeneration/adgeneration.go

```diff
@@ -0,0 +1,262 @@
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type AdgenerationAdapter struct {
+	endpoint        string
+	version         string
+	defaultCurrency string
+}
+
+// Server Responses
+type adgServerResponse struct {
+	Locationid string        `json:"locationid"`
+	Dealid     string        `json:"dealid"`
+	Ad         string        `json:"ad"`
+	Beacon     string        `json:"beacon"`
+	Beaconurl  string        `json:"beaconurl"`
+	Cpm        float64       `jsons:"cpm"`
+	Creativeid string        `json:"creativeid"`
+	H          uint64        `json:"h"`
+	W          uint64        `json:"w"`
+	Ttl        uint64        `json:"ttl"`
+	Vastxml    string        `json:"vastxml,omitempty"`
+	LandingUrl string        `json:"landing_url"`
+	Scheduleid string        `json:"scheduleid"`
+	Results    []interface{} `json:"results"`
+}
+
+func (adg *AdgenerationAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	numRequests := len(request.Imp)
+	var errs []error
+
+	if numRequests == 0 {
+		errs = append(errs, &errortypes.BadInput{
+			Message: "No impression in the bid request",
+		})
+		return nil, errs
+	}
+
+	requestArray := make([]*adapters.RequestData, 0, numRequests)
+	for index := 0; index < numRequests; index++ {
+		headers := http.Header{}
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+		headers.Add("Accept", "application/json")
+
+		requestUri, err := adg.getRequestUri(request, index)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+		request := &adapters.RequestData{
+			Method:  "GET",
+			Uri:     requestUri,
+			Body:    nil,
+			Headers: headers,
+		}
+		requestArray = append(requestArray, request)
+	}
+
+	return requestArray, errs
+}
+
+func (adg *AdgenerationAdapter) getRequestUri(request *openrtb.BidRequest, index int) (string, error) {
+	imp := request.Imp[index]
+	bidderExt, err := adg.unmarshalExtImpBidder(&imp)
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	adgExt, err := adg.unmarshalExtImpAdgeneration(&bidderExt)
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	endpoint := adg.endpoint +
+		"?posall=SSPLOC" +
+		"&id=" + adgExt.Id +
+		"&sdktype=0" +
+		"&hb=true" +
+		"&t=json3" +
+		"&currency=" + adg.getCurrency(request) +
+		"&sdkname=prebidserver" +
+		"&adapterver=" + adg.version
+	if adg.getSizes(&imp) != "" {
+		endpoint += "&sizes=" + adg.getSizes(&imp)
+	}
+	if request.Site != nil && request.Site.Page != "" {
+		endpoint += "&tp=" + request.Site.Page
+	}
+
+	return endpoint, nil
+}
+
+func (adg *AdgenerationAdapter) unmarshalExtImpBidder(imp *openrtb.Imp) (adapters.ExtImpBidder, error) {
```

Comment:
Many of these methods do not reference the `AdgenerationAdapter` object, so does not need to be a method of the object. Was there any reason you made them methods rather than simple helper functions?